The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11172025942/job/31057827238) [(plain)](https://github.com/rust-lang/rust/commit/c52a0f8f2bb8677f8b8fd6b0e9ba6571eab73975/checks/31057827238/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20950f2ccee3dff53a038adf5c1cf05231c0b30772617126a5f6478a66316a29cba9aead69f7bb0004886d32c1f8e6287542cf6d25130711c82d16a66201d4fe:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.57s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [rustdoc] tests/rustdoc/intra-doc/basic.rs stdout ----

error: rustdoc failed!
status: exit status: 1
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/intra-doc/basic" && env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/intra-doc/basic/auxiliary" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/intra-doc/basic" "--deny" "warnings" "/checkout/tests/rustdoc/intra-doc/basic.rs" "-A" "internal_features"
--- stderr -------------------------------
error: `ThisEnum::ThisVariant` is both a unit variant and a variant
##[error]  --> /checkout/tests/rustdoc/intra-doc/basic.rs:48:33
   |
   |
48 | //! * [`ThisEnum::ThisVariant`](ThisEnum::ThisVariant)
   |                                 ^^^^^^^^^^^^^^^^^^^^^ ambiguous link
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
help: to link to the unit variant, prefix with `variant@`
   |
   |
48 | //! * [`ThisEnum::ThisVariant`](variant@ThisEnum::ThisVariant)
help: to link to the variant, prefix with `variant@`
   |
   |
48 | //! * [`ThisEnum::ThisVariant`](variant@ThisEnum::ThisVariant)

error: `ThisEnum::ThisVariantCtor` is both a tuple variant and a variant
##[error]  --> /checkout/tests/rustdoc/intra-doc/basic.rs:49:37
   |
   |
49 | //! * [`ThisEnum::ThisVariantCtor`](ThisEnum::ThisVariantCtor)
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous link
   |
help: to link to the tuple variant, prefix with `variant@`
   |
49 | //! * [`ThisEnum::ThisVariantCtor`](variant@ThisEnum::ThisVariantCtor)
help: to link to the variant, prefix with `variant@`
   |
   |
49 | //! * [`ThisEnum::ThisVariantCtor`](variant@ThisEnum::ThisVariantCtor)

error: `ThisEnum::ThisVariant` is both a unit variant and a variant
##[error]   --> /checkout/tests/rustdoc/intra-doc/basic.rs:110:8
    |
    |
110 | /// * [ThisEnum::ThisVariant]
    |        ^^^^^^^^^^^^^^^^^^^^^ ambiguous link
help: to link to the unit variant, prefix with `variant@`
    |
    |
110 | /// * [variant@ThisEnum::ThisVariant]
help: to link to the variant, prefix with `variant@`
    |
    |
110 | /// * [variant@ThisEnum::ThisVariant]

error: aborting due to 3 previous errors
------------------------------------------



---- [rustdoc] tests/rustdoc/intra-doc/field.rs stdout ----

error: rustdoc failed!
status: exit status: 1
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/intra-doc/field" && env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/intra-doc/field/auxiliary" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/intra-doc/field" "--deny" "warnings" "/checkout/tests/rustdoc/intra-doc/field.rs" "-A" "internal_features"
--- stderr -------------------------------
error: `std::io::ErrorKind::NotFound` is both a unit variant and a variant
##[error] --> /checkout/tests/rustdoc/intra-doc/field.rs:6:17
  |
  |
6 | //! [not_found][std::io::ErrorKind::NotFound]
  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous link
  = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
help: to link to the unit variant, prefix with `variant@`
  |
  |
6 | //! [not_found][variant@std::io::ErrorKind::NotFound]
help: to link to the variant, prefix with `variant@`
  |
  |
6 | //! [not_found][variant@std::io::ErrorKind::NotFound]

error: aborting due to 1 previous error
------------------------------------------



---- [rustdoc] tests/rustdoc/struct-field.rs stdout ----

error: rustdoc failed!
status: exit status: 1
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/struct-field" && env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/struct-field/auxiliary" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/struct-field" "--deny" "warnings" "/checkout/tests/rustdoc/struct-field.rs" "-A" "internal_features"
--- stderr -------------------------------
--- stderr -------------------------------
error: `Uniooon::X` is both a unit variant and a variant
  |
  |
8 | //! Test with [Foo::bar], [Bar::foo], [Uniooon::X]
  |                                        ^^^^^^^^^^ ambiguous link
  = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
help: to link to the unit variant, prefix with `variant@`
  |
  |
8 | //! Test with [Foo::bar], [Bar::foo], [variant@Uniooon::X]
help: to link to the variant, prefix with `variant@`
  |
  |
8 | //! Test with [Foo::bar], [Bar::foo], [variant@Uniooon::X]

error: aborting due to 1 previous error
------------------------------------------

```

</details>

--- a/compiler/rustc_resolve/src/lib.rs
+++ b/compiler/rustc_resolve/src/lib.rs
@@ -2148,7 +2148,9 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
 
         match self.maybe_resolve_path(&segments, Some(ns), &parent_scope, None) {
             PathResult::Module(ModuleOrUniformRoot::Module(module)) => Some(module.res().unwrap()),
-            PathResult::NonModule(path_res) => path_res.full_res(),
+            PathResult::NonModule(path_res) => {
+                path_res.full_res().filter(|res| !matches!(res, Res::Def(DefKind::Ctor(..), _)))
+            }
             PathResult::Module(ModuleOrUniformRoot::ExternPrelude) | PathResult::Failed { .. } => {
                 None
             }

--- a/tests/rustdoc-ui/intra-doc/disambiguator-mismatch.rs
+++ b/tests/rustdoc-ui/intra-doc/disambiguator-mismatch.rs
@@ -91,7 +91,7 @@ struct X {
 //~| HELP prefix with `field@`
 
 /// Link to [field@S::A]
-//~^ ERROR incompatible link kind for `S::A`
-//~| NOTE this link resolved
+//~^ ERROR unresolved link to `S::A`
+//~| NOTE this link resolves
 //~| HELP prefix with `variant@`
 pub fn f() {}

--- a/tests/rustdoc-ui/intra-doc/disambiguator-mismatch.stderr
+++ b/tests/rustdoc-ui/intra-doc/disambiguator-mismatch.stderr
@@ -160,13 +160,13 @@ help: to link to the field, prefix with `field@`
 LL | /// Link to [field@X::y]
    |              ~~~~~~
 
-error: incompatible link kind for `S::A`
+error: unresolved link to `S::A`
   --> $DIR/disambiguator-mismatch.rs:93:14
    |
 LL | /// Link to [field@S::A]
-   |              ^^^^^^^^^^ this link resolved to a unit variant, which is not a field
+   |              ^^^^^^^^^^ this link resolves to the variant `A`, which is not in the value namespace
    |
-help: to link to the unit variant, prefix with `variant@`
+help: to link to the variant, prefix with `variant@`
    |
 LL | /// Link to [variant@S::A]
    |              ~~~~~~~~

--- a/tests/rustdoc-ui/intra-doc/value-ctor.rs
+++ b/tests/rustdoc-ui/intra-doc/value-ctor.rs
@@ -1,0 +1,35 @@
+// https://github.com/rust-lang/rust/issues/130591
+#![deny(rustdoc::broken_intra_doc_links)]
+#![crate_name = "foo"]
+
+/// [value@Foo::X] //~ERROR broken
+pub enum Foo {
+    X,
+}
+
+/// [tst][value@MyStruct] //~ERROR broken
+pub struct MyStruct;
+
+pub enum MyEnum {
+    Internals,
+}
+
+pub use MyEnum::*;
+
+/// In this context, [a][type@Internals] is a struct,
+/// while [b][value@Internals] fails. //~ERROR broken
+/// Also, [c][struct@Internals] is a struct,
+/// while [d][variant@Internals] fails. //~ERROR broken
+pub struct Internals {
+    foo: (),
+}
+
+pub mod inside {
+    pub struct Internals2;
+}
+
+use inside::*;
+
+/// In this context, [a][type@Internals2] is an enum,
+/// while [b][value@Internals2] fails. //~ERROR broken
+pub enum Internals2 {}

--- a/tests/rustdoc-ui/intra-doc/value-ctor.stderr
+++ b/tests/rustdoc-ui/intra-doc/value-ctor.stderr
@@ -1,0 +1,62 @@
+error: unresolved link to `Foo::X`
+  --> $DIR/value-ctor.rs:5:6
+   |
+LL | /// [value@Foo::X]
+   |      ^^^^^^^^^^^^ this link resolves to the variant `X`, which is not in the value namespace
+   |
+note: the lint level is defined here
+  --> $DIR/value-ctor.rs:2:9
+   |
+LL | #![deny(rustdoc::broken_intra_doc_links)]
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+help: to link to the variant, prefix with `variant@`
+   |
+LL | /// [variant@Foo::X]
+   |      ~~~~~~~~
+
+error: unresolved link to `MyStruct`
+  --> $DIR/value-ctor.rs:10:11
+   |
+LL | /// [tst][value@MyStruct]
+   |           ^^^^^^^^^^^^^^ this link resolves to the struct `MyStruct`, which is not in the value namespace
+   |
+help: to link to the struct, prefix with `struct@`
+   |
+LL | /// [tst][struct@MyStruct]
+   |           ~~~~~~~
+
+error: unresolved link to `Internals`
+  --> $DIR/value-ctor.rs:20:15
+   |
+LL | /// while [b][value@Internals] fails.
+   |               ^^^^^^^^^^^^^^^ this link resolves to the struct `Internals`, which is not in the value namespace
+   |
+help: to link to the struct, prefix with `struct@`
+   |
+LL | /// while [b][struct@Internals] fails.
+   |               ~~~~~~~
+
+error: incompatible link kind for `Internals`
+  --> $DIR/value-ctor.rs:22:15
+   |
+LL | /// while [d][variant@Internals] fails.
+   |               ^^^^^^^^^^^^^^^^^ this link resolved to a struct, which is not a variant
+   |
+help: to link to the struct, prefix with `struct@`
+   |
+LL | /// while [d][struct@Internals] fails.
+   |               ~~~~~~~
+
+error: unresolved link to `Internals2`
+  --> $DIR/value-ctor.rs:34:15
+   |
+LL | /// while [b][value@Internals2] fails.
+   |               ^^^^^^^^^^^^^^^^ this link resolves to the enum `Internals2`, which is not in the value namespace
+   |
+help: to link to the enum, prefix with `enum@`
+   |
+LL | /// while [b][enum@Internals2] fails.
+   |               ~~~~~
+
+error: aborting due to 5 previous errors
+

